libraries/chain/kv_context_chainbase.cpp

```diff
@@ -0,0 +1,225 @@
+      }
+
+      int32_t kv_it_key_compare(const char* key, uint32_t size) override {
+         if (!kv_key)
+            return 1;
+         return compare_blob(*kv_key, std::string_view{ key, size });
+      }
+
+      kv_it_stat kv_it_move_to_end() override {
+         kv_key.reset();
+         return kv_it_stat::iterator_end;
+      }
+
+      kv_it_stat kv_it_next() override {
+         if (kv_key) {
+            auto it = idx.lower_bound(boost::make_tuple(database_id, contract, *kv_key));
+            if (it != idx.end()) {
+               if (it->kv_key == *kv_key)
+                  ++it;
+               return move_to(it);
+            }
+            kv_key.reset();
+            return kv_it_stat::iterator_end;
+         }
+         return move_to(idx.lower_bound(boost::make_tuple(database_id, contract, prefix)));
+      }
+
+      kv_it_stat kv_it_prev() override {
+         std::decay_t<decltype(idx.end())> it;
+         if (kv_key)
+            it = idx.lower_bound(boost::make_tuple(database_id, contract, *kv_key));
+         else if (next_prefix)
+            it = idx.lower_bound(boost::make_tuple(database_id, contract, *next_prefix));
+         else
+            it = idx.end();
+         if (it != idx.begin())
+            return move_to(--it);
+         kv_key.reset();
+         return kv_it_stat::iterator_end;
+      }
+
+      kv_it_stat kv_it_lower_bound(const char* key, uint32_t size) override {
+         return move_to(idx.lower_bound(boost::make_tuple(database_id, contract, std::string_view{ key, size })));
+      }
+
+      kv_it_stat kv_it_key(uint32_t offset, char* dest, uint32_t size, uint32_t& actual_size) override {
+         if (kv_key) {
+            if (offset < kv_key->size())
+               memcpy(dest, kv_key->data() + offset, std::min((size_t)size, kv_key->size() - offset));
+            actual_size = kv_key->size();
+            return kv_it_status();
+         } else {
+            actual_size = 0;
+            return kv_it_stat::iterator_end;
+         }
+      }
+
+      kv_it_stat kv_it_value(uint32_t offset, char* dest, uint32_t size, uint32_t& actual_size) override {
+         if (!kv_key) {
+            actual_size = 0;
+            return kv_it_stat::iterator_end;
+         }
+         auto* kv = db.find<kv_object, by_kv_key>(boost::make_tuple(database_id, contract, *kv_key));
+         if (!kv) {
+            actual_size = 0;
+            return kv_it_stat::iterator_erased;
+         }
+         if (offset < kv->kv_value.size())
+            memcpy(dest, kv->kv_value.data() + offset, std::min((size_t)size, kv->kv_value.size() - offset));
+         actual_size = kv->kv_value.size();
+         return kv_it_stat::iterator_ok;
+      }
+   }; // kv_iterator_chainbase
+
+   struct kv_context_chainbase : kv_context {
+      chainbase::database&       db;
+      name                       database_id;
+      name                       receiver;
+      std::optional<shared_blob> temp_data_buffer;
```

Comment:
I think optional is pointless.  There's no difference in behavior between an empty buffer and a buffer that holds an empty string, right.

Reply:
It prevents me from having to construct a shared_blob, which needs an allocator